pennylane/transforms/decompose.py

```diff
@@ -0,0 +1,210 @@
+        3: ────╰QuantumPhaseEstimation─┤
+
+    >>> print(qml.draw(decompose(circuit, max_expansion=1))())
+
+    .. code-block:: python3
+
+        0: ──H─╭U(M0)⁴─╭U(M0)²─╭U(M0)¹───────┤
+        1: ──H─╰●──────│───────│───────╭QFT†─┤
+        2: ──H─────────╰●──────│───────├QFT†─┤
+        3: ──H─────────────────╰●──────╰QFT†─┤
+
+    >>> print(qml.draw(decompose(circuit, max_expansion=2))())
+
+    .. code-block:: python3
+
+        0: ──H──RZ(11.00)──RY(1.14)─╭X──RY(-1.14)──RZ(-9.42)─╭X──RZ(-1.57)──RZ(1.57)──RY(1.00)─╭X──RY(-1.00)
+        1: ──H──────────────────────╰●───────────────────────╰●────────────────────────────────│────────────
+        2: ──H─────────────────────────────────────────────────────────────────────────────────╰●───────────
+        3: ──H──────────────────────────────────────────────────────────────────────────────────────────────
+
+        ───RZ(-6.28)─╭X──RZ(4.71)──RZ(1.57)──RY(0.50)─╭X──RY(-0.50)──RZ(-6.28)─╭X──RZ(4.71)─────────────────
+        ─────────────│────────────────────────────────│────────────────────────│──╭SWAP†────────────────────
+        ─────────────╰●───────────────────────────────│────────────────────────│──│─────────────╭(Rϕ(1.57))†
+        ──────────────────────────────────────────────╰●───────────────────────╰●─╰SWAP†─────H†─╰●──────────
+
+        ────────────────────────────────────┤
+        ──────╭(Rϕ(0.79))†─╭(Rϕ(1.57))†──H†─┤
+        ───H†─│────────────╰●───────────────┤
+        ──────╰●────────────────────────────┤
+    """
+
+    if gate_set is None:
+        gate_set = set(qml.ops.__all__)
+
+    if isinstance(gate_set, (str, type)):
+        gate_set = set([gate_set])
+
+    if isinstance(gate_set, Iterable):
+        gate_types = tuple(gate for gate in gate_set if isinstance(gate, type))
+        gate_names = set(gate for gate in gate_set if isinstance(gate, str))
+        gate_set = lambda op: (op.name in gate_names) or isinstance(op, gate_types)
+
+    def decomposer(op):
+        return op.decomposition()
+
+    def stopping_condition(op):
+        if not op.has_decomposition:
+            if not gate_set(op):
+                warnings.warn(
+                    f"Operator {op.name} has no supported decomposition and was not found in the set of allowed decomposition gates."
+                    f"To remove this warning, add the operator name ({op.name}) or type ({type(op)}) to the allowed set of gates.",
+                    UserWarning,
+                )
+            return True
+        return gate_set(op)
+
+    if all(stopping_condition(op) for op in tape.operations):
+        return (tape,), null_postprocessing
+
+    try:
+        new_ops = [
+            final_op
+            for op in tape.operations
+            for final_op in qml.devices.preprocess._operator_decomposition_gen(
+                op,
+                stopping_condition,
+                decomposer,
+                max_expansion=max_expansion,
+            )
+        ]
```

Comment:
This might be a case where we actually do want some code duplication, but it is borderline.  

By repurposing that code, we might not need to change two places if something forces us to change that method.  But if we need to change that method for one use case and not the other? Then we have to duplicate the method and disentangle the dependency at that time. Especially since future developers probably won't be aware this module is depending on that private method, they may break things inadvertently.

I'm fine with keeping the coupling between the two versions, but we may want to add a comment to `_operator_decomposition_gen` that it is also used here and to be careful about changing it. I would also consider just copying over some of the code though.